README.md
# Hey fellow Rails developers, please read!

**Want to use rack-zippy with a Rails v4.2 or greater app?**  
Its recommended you don't! Rails 4.2+ now supports serving gzipped files directly so there's no need for rack-zippy in Rails 4.2+ apps.

**Want to use rack-zippy with a Rails v4.1 or less app?**  
You'll need to use v3.0 of rack-zippy, see the README here: https://github.com/eliotsykes/rack-zippy/tree/v3.0.1

# rack-zippy

rack-zippy v4+ is a Rack middleware for serving .gz files in Rack apps that are **not** Rails 4.2+ apps. (If you need to use rack-zippy in a Rails <= 4.1 app, then use v3.0 of rack-zippy, see README here: https://github.com/eliotsykes/rack-zippy/tree/v3.0.1)

rack-zippy has convenient directory request handling:

- Requests for `/` and `/index` respond with `public/index.html` if present
- Requests for `/foo/` and `/foo` respond with first file present out of `public/foo.html`, `public/foo/index.html` (Same behaviour for subdirectories)

rack-zippy decorates actionpack's `ActionDispatch::Static` middleware for non-Rails Rack apps to provide rack-zippy's own choice of caching headers and whitelisting of permitted static file extensions. (As an alternative to rack-zippy, you can use actionpack's `ActionDispatch::Static` directly without rack-zippy.)

## Installation in Rack app (that isn’t a Rails app)

Add this line to your application's Gemfile:

    gem 'rack-zippy'

And then execute:

    $ bundle

In `config.ru`:

    require 'rack-zippy'

    # Set asset_root to an absolute or relative path to the directory holding your asset files
    # e.g. '/path/to/my/apps/static-assets' or 'public'
    asset_root = '/path/to/my/apps/public'
    use Rack::Zippy::AssetServer, asset_root


## Usage

Follow the installation instructions above and rack-zippy will serve any static assets, including gzipped assets, from your
application's public/ directory and will respond with sensible caching headers.

### Options

#### max_age_fallback

`max_age_fallback`, is an integer value in seconds that should be used as the max_age fallback for files served by rack-zippy that live **outside** the `/assets` subdirectory *and* aren't `/favicon.ico`.

A typical use for `max_age_fallback` is to define how long the cache lifetime for static HTML files served by rack-zippy should be. For one of my sites I have this set to 15 minutes:

```ruby
max_age_in_secs = 15*60 # 15 mins = 900 secs
use Rack::Zippy::AssetServer, asset_root, max_age_fallback: max_age_in_secs
```

Any files given the `max_age_fallback` would have the following `Cache-Control` header:

```
Cache-Control: public, max-age=900
```

### Configuration

#### Supported Extensions Whitelist

rack-zippy handles only files with whitelisted extensions. Default extensions are stored in the `static_extensions` array with an entry for each of these:
`css js html htm txt ico png jpg jpeg gif pdf svg zip gz eps psd ai woff woff2 ttf eot otf swf`

You can modify this list to support other extensions by appending the lowercased file extension to the `static_extensions` array:

```ruby
Rack::Zippy.configure do |config|
  # Add support for the given extensions:
  config.static_extensions.push('map', 'csv', 'xls', 'rtf', ...EXTENSIONS TO ADD...)
end
```

It is not recommended, however if you use rack-zippy 4.0+ with a Rails 4.2+ app, you can skip the rack-zippy rails version check and log output. Put the following in an initializer:

```ruby
# config/initializers/zippy.rb
Rack::Zippy::Railtie.skip_version_check = true
```


## Troubleshooting

##### NameError: uninitialized constant Rack::Zippy

- Check `Gemfile` doesn't limit rack-zippy to a subset of environment groups
- Run `bundle install`
- Check `Gemfile.lock` contains an entry for rack-zippy
- Ensure `require 'rack-zippy'` is present near the top of `config.ru`


## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Run tests (`rake test`)
4. Commit your changes (`git commit -am 'Add some feature'`)
5. Push to the branch (`git push origin my-new-feature`)
6. Create new Pull Request

### Optional for contributors
To try a local branch of rack-zippy out as the gem dependency in a local app, configure bundler with a local gem
override as follows:

In `your-app/Gemfile`: edit the rack-zippy dependency to the following:

    # The branch your-local-branch-name **must** exist otherwise bundler will shout obscenities at you
    gem 'rack-zippy', :github => 'eliotsykes/rack-zippy', :branch => 'your-local-branch-name'

At the command line, inside `your-app`, configure bundler to set a local git repo to override the one we specified in the previous step for rack-zippy:

    $> bundle config --local local.rack-zippy /path/to/your/local/rack-zippy

Now when you run your-app **with** `bundle exec`, the rack-zippy gem dependency will resolve to `/path/to/your/local/rack-zippy`.

Cleanup time! When you’re finished testing, delete the local override and set your Gemfile dependency back to the original:

    # At the command line:
    $> bundle config --delete local.rack-zippy

    # In your-app/Gemfile change rack-zippy dependency to this (or similar):
    gem 'rack-zippy', '~> 9.8.7' # Replace 9.8.7 with the rack-zippy release version you want to use.

#### How to Run a Single Test

```bash
# Single test file
ruby -Ilib:test test/assert_server_test.rb

# Single test method
ruby -Ilib:test test/assert_server_test.rb --name test_serves_static_file_as_directory

# Test methods matching a regex
ruby -Ilib:test test/assert_server_test.rb --name /serves_static/
```

## Contributors

- Eliot Sykes https://eliotsykes.com
- Kieran Topping https://github.com/ktopping
- Luke Wendling https://github.com/lukewendling
- Anton Petrunich https://github.com/solenko
- ssemakov https://github.com/ssemakov
- Kai Schlichting https://github.com/lacco

## Releasing a new gem

1. Update pre-release version to the release version in `lib/rack-zippy/version.rb`, e.g. `1.0.1.pre` becomes `1.0.1`
2. Update `CHANGELOG.md` version and date. Update Contributors in `README.md`.
3. Tests pass? (`rake test`)
4. Commit and push changes to origin.
4. Build the gem (`rake build`)
5. Release on rubygems.org (`rake release`)
6. Update version to the next pre-release version in `lib/rack-zippy/version.rb`, e.g. `1.0.1` becomes `1.0.2.pre`.
7. Add new heading to `CHANGELOG` for the next pre-release
8. Commit and push the updated `lib/rack-zippy/version.rb` and `CHANGELOG` files.


CHANGELOG.md
## 4.0.1.pre / yyyy-mm-dd

## 4.0.0 / 2015-10-20
- Rack Zippy continues to support non-Rails Rack apps, though Rails support has ended in rack-zippy 4.0+. Rails 4.2+ now supports serving gzipped files directly. Configure your Rails 4.2+ app to use the `ActionDispatch::Static` middleware. If you absolutely need rack-zippy for your Rails app or can't upgrade to Rails 4.2+, try using an earlier rack-zippy version: `~> 3.0.1`
- New dependency on actionpack gem. Rack Zippy decorates the `ActionDispatch::Static` middleware for non-Rails Rack apps to provide Rack Zippy's own choice of caching headers and whitelisting of permitted static file extensions.

## 3.0.1 / 2015-05-19
- Allow paths with periods in the middle, but not if they immediately follow slash ([#46](https://github.com/eliotsykes/rack-zippy/pull/46)) @ssemakov

## 3.0.0 / 2015-04-18
- POTENTIAL BREAKING CHANGE! `STATIC_EXTENSION_REGEX` has been removed and replaced with a `static_extensions` array. If your app monkey patched `STATIC_EXTENSTION_REGEX` to change the file extension whitelist, then you will need to update your app to use rack-zippy 3.x. Depending on how you patched, and your test coverage, your app could silently fail. Search your codebase for `STATIC_EXTENSION_REGEX` to ensure it is not used. If it is used, then migrate your patch to use the new `Rack::Zippy.config` method for configuring `static_extensions`: https://github.com/eliotsykes/rack-zippy#static_extensions
- Make static extensions list configurable ([#45](https://github.com/eliotsykes/rack-zippy/pull/45)) Anton Petrunich

## 2.0.2 / 2014-12-15
- Remove binstub bin/rake permanently, may be causing Heroku issues

## 2.0.1 / 2014-12-15
- New `:max_age_fallback` (in seconds) option

## 2.0.0 / 2014-12-07
- Rack Zippy now works cleanly with all Rack apps, including Rails apps ([#23](https://github.com/eliotsykes/rack-zippy/issues/23))
- Decomposed AssetServer into AssetServer, ServeableFile, and AssetCompiler classes
- Installation notes for non-Rails Rack apps added to README.md
- Smarter handling of directory requests to match behaviour of Rails Static middleware ([#15](https://github.com/eliotsykes/rack-zippy/issues/15))
    - Requests for `/` and `/index` respond with `public/index.html` if present
    - Requests for `/foo/` and `/foo` respond with first file present out of `public/foo.html`, `public/foo/index.html` (Same behaviour for subdirectories)
- Use File.join to build file path ([#34](https://github.com/eliotsykes/rack-zippy/issues/34))
- Respond with 404 Not Found instead of raising SecurityError for illegal paths ([#17](https://github.com/eliotsykes/rack-zippy/issues/17))

## 1.2.1 / 2014-07-09
- Use absolute (not relative) path for default asset_root ([#11](https://github.com/eliotsykes/rack-zippy/issues/11))

## 1.2.0 / 2014-07-09
- Add handling of font files with extensions: woff, woff2, ttf, eot, otf
  ([#9](https://github.com/eliotsykes/rack-zippy/issues/9))
- Handle flash .swf file extension ([#20](https://github.com/eliotsykes/rack-zippy/issues/20))

## 1.1.0 / 2014-01-26
- rack-zippy no longer blocks requests for assets that it cannot find on the filesystem. These
  requests are now passed on to the app. ([#7](https://github.com/eliotsykes/rack-zippy/issues/7))

## 1.0.1 / 2013-09-06
-  Fix error on request for assets root dir (fixes [#2](https://github.com/eliotsykes/rack-zippy/issues/2))
-  Add MIT license to gemspec (fixes [#1](https://github.com/eliotsykes/rack-zippy/issues/1))

## 1.0.0 / 2013-08-22

Production-ready release

-  Only set years-old cache-friendly last-modified header on assets with max-age of at least 1 month

## 0.1.0 / 2013-08-21

Initial release


test/test_helper.rb
require 'rubygems'
require 'bundler/setup'

Bundler.require :default, :development

require 'minitest/autorun'
require 'rack/test'
require 'active_support/testing/declarative'


class TestCase < ::Minitest::Test
  extend ActiveSupport::Testing::Declarative

  DURATIONS_IN_SECS = {:year => 31536000, :month => 2678400, :day => 86400}.freeze

  def ensure_correct_working_directory
    is_project_root_working_directory = ::File.exists?('rack-zippy.gemspec')
    if is_project_root_working_directory
      @original_dir = Dir.pwd
      Dir.chdir 'test'
    end
  end

  def revert_to_original_working_directory
    Dir.chdir @original_dir if @original_dir
  end

  def asset_root
    "#{Dir.pwd}/public"
  end


end


test/railtie_test.rb
require_relative 'test_helper'

module Rails
  class Railtie
    def self.config
      OpenStruct.new(after_initialize: true)
    end
  end

  def self.version
    raise 'Stub Rails.version'
  end
end

require 'rack-zippy/railtie'

module Rack
  module Zippy
    class RailtieTest <  TestCase

      def teardown
        Rack::Zippy::Railtie.skip_version_check = false
      end

      def test_upgrade_rails_or_downgrade_zippy_recommendation_for_rails_versions_below_4_2
        with_rails_version('4.1.99') do
          assert_output("[rack-zippy] This version of rack-zippy does not support Rails. Your choices include: 1) [RECOMMENDED] Upgrade to Rails 4.2 or above and use Rails' built-in ActionDispatch::Static middleware to serve gzipped files. or 2) Specify an earlier version of rack-zippy (~> '3.0.1') in your Gemfile that does support Rails\n") { Rack::Zippy::Railtie.version_check }
        end
      end

      versions = ['4.2.0', '5.0.0', '10.0.0']
      versions.each do |version|
        test "remove zippy recommendation printed for Rails versions #{version}" do
            with_rails_version(version) do
              assert_output("[rack-zippy] rack-zippy is not supported for this version of Rails. Rails now supports serving gzipped files using its own ActionDispatch::Static middleware. It is strongly recommended you remove rack-zippy from your app and use ActionDispatch::Static in its place.\n") { Rack::Zippy::Railtie.version_check }
            end
        end
      end

      def test_config_silences_remove_zippy_recommendation_for_rails_versions_4_2_and_above
        versions = ['4.2.0', '5.0.0', '10.0.0']
        versions.each do |version|
          with_rails_version(version) do
            Rack::Zippy::Railtie.skip_version_check = true
            assert_silent { Rack::Zippy::Railtie.version_check }
          end
        end
      end

      private

      def with_rails_version(version)
        Rails.stub :version, version do
          yield
        end
      end

    end
  end
end


test/version_test.rb
require_relative 'test_helper'
require 'minitest/autorun'
require 'rack-zippy/version'

module Rack
  module Zippy
    class VersionTest <  Minitest::Test

      def test_no_remove_zippy_recommendation_for_non_rails_apps
        assert_silent { Rack::Zippy.version_check }
      end

      def test_no_remove_zippy_recommendation_for_rails_versions_below_4_2
        assert_silent { version_check }
        stub Rails.version with "4.1.99"
        flunk
      end

      def test_remove_zippy_recommendation_for_rails_versions_4_2_and_above
        stub Rails.version with "4.2.0", "5.0.0"
        flunk
      end

      def test_config_silences_remove_zippy_recommendation_for_rails_versions_4_2_and_above
        # Use Rack::Zippy.configuration
        flunk
      end

    end
  end
end


lib/rack-zippy-old.rb
require 'rack-zippy/version'
require 'rack-zippy/asset_compiler'
require 'rack-zippy/serveable_file'
require 'rack-zippy/configuration'

module Rack
  module Zippy
    extend Configuration

    define_setting :static_extensions, %w(css js html htm txt ico png jpg jpeg gif pdf svg zip gz eps psd ai woff woff2 ttf eot otf swf)

    PRECOMPILED_ASSETS_SUBDIR_REGEX = /\A\/assets(?:\/|\z)/

    class AssetServer

      HTTP_STATUS_CODE_OK = 200

      def initialize(app, asset_root=nil, options={})
        raise ArgumentError, 'options argument cannot be a String' if options.is_a? String
        if asset_root.nil?
          if RailsAssetCompiler.rails_env?
            asset_root = ::Rails.public_path
          else
            raise ArgumentError, 'Please specify asset_root when initializing Rack::Zippy::AssetServer ' +
              '(asset_root is the path to your public directory, often the one with favicon.ico in it)'
          end
        end
        @app = app

        @options = {
          :asset_compiler => resolve_asset_compiler,
          :asset_root => asset_root
        }.merge(options)
      end

      def call(env)
        path_info = env['PATH_INFO']

        return not_found_response if path_info =~ ILLEGAL_PATH_REGEX

        serveable_file_options = {
            :path_info => path_info,
            :asset_root => asset_root,
            :asset_compiler => asset_compiler,
            :include_gzipped => client_accepts_gzip?(env),
            :max_age_fallback => @options[:max_age_fallback]
        }

        serveable_file = ServeableFile.find_first(serveable_file_options)


        if serveable_file
          return [HTTP_STATUS_CODE_OK, serveable_file.headers, serveable_file.response_body]
        end

        @app.call(env)
      end

      def asset_root
        @options[:asset_root]
      end

      private

      ACCEPTS_GZIP_REGEX = /\bgzip\b/

      ILLEGAL_PATH_REGEX = /(\/\.\.?)/

      def client_accepts_gzip?(rack_env)
        rack_env['HTTP_ACCEPT_ENCODING'] =~ ACCEPTS_GZIP_REGEX
      end

      def resolve_asset_compiler
        asset_compiler_class = RailsAssetCompiler.rails_env? ? RailsAssetCompiler : NullAssetCompiler
        return asset_compiler_class.new
      end

      def asset_compiler
        @options[:asset_compiler]
      end

      def not_found_response
        [404, {}, ['Not Found']]
      end

    end
  end
end


lib/rack-zippy/railtie.rb
module Rack
  module Zippy
    class Railtie < ::Rails::Railtie

      cattr_accessor :skip_version_check

      def self.version_check
        return if Rack::Zippy::Railtie.skip_version_check
        if ::Rails.version.to_f >= minimum_rails_version_with_gzip_serving
          puts "[rack-zippy] rack-zippy is not supported for this version of Rails. Rails now supports serving gzipped files using its own ActionDispatch::Static middleware. It is strongly recommended you remove rack-zippy from your app and use ActionDispatch::Static in its place."
        else
          puts "[rack-zippy] This version of rack-zippy does not support Rails. Your choices include: 1) [RECOMMENDED] Upgrade to Rails #{minimum_rails_version_with_gzip_serving} or above and use Rails' built-in ActionDispatch::Static middleware to serve gzipped files. or 2) Specify an earlier version of rack-zippy (~> '3.0.1') in your Gemfile that does support Rails"
        end
      end

      def self.minimum_rails_version_with_gzip_serving
        4.2
      end

      config.after_initialize do
        Rack::Zippy::Railtie.version_check
      end

    end
  end
end


lib/rack-zippy/configuration.rb
module Rack
  module Zippy
    module Configuration
      def configure
        yield self
      end

      def define_setting(name, default = nil)
        default_copy = begin
                         default.dup
        rescue TypeError
                         default
        end
        class_variable_set("@@#{name}", default_copy)

        define_class_method "#{name}=" do |value|
          class_variable_set("@@#{name}", value)
        end
        define_class_method name do
          class_variable_get("@@#{name}")
        end
        define_class_method "reset_#{name}" do
          class_variable_set("@@#{name}", default)
        end
      end

      private

      def define_class_method(name, &block)
        (class << self; self; end).instance_eval do
          define_method name, &block
        end
      end

    end
  end
end


lib/rack-zippy/serveable_file.rb
module Rack
  module Zippy
    class ServeableFile

      attr_reader :path, :full_path_info

      def initialize(options)
        raise ArgumentError.new(':has_encoding_variants option must be given') unless options.has_key?(:has_encoding_variants)

        @path = options[:path]
        @full_path_info = options[:full_path_info]
        @has_encoding_variants = options[:has_encoding_variants]
        @is_gzipped = options[:is_gzipped]
        @max_age_fallback = options[:max_age_fallback] || SECONDS_IN[:day]
      end

      def headers
        headers = { 'Content-Type'  => Rack::Mime.mime_type(::File.extname(full_path_info)) }
        headers.merge! cache_headers

        headers['Vary'] = 'Accept-Encoding' if encoding_variants?
        headers['Content-Encoding'] = 'gzip' if gzipped?

        headers['Content-Length'] = ::File.size(path).to_s
        return headers
      end

      def cache_headers
        case full_path_info
        when PRECOMPILED_ASSETS_SUBDIR_REGEX
          lifetime_in_secs = SECONDS_IN[:year]
          last_modified = CACHE_FRIENDLY_LAST_MODIFIED
        when '/favicon.ico'
          lifetime_in_secs = SECONDS_IN[:month]
          last_modified = CACHE_FRIENDLY_LAST_MODIFIED
        else
          lifetime_in_secs = @max_age_fallback
        end

        headers = { 'Cache-Control' => "public, max-age=#{lifetime_in_secs}" }
        headers['Last-Modified'] = last_modified if last_modified

        return headers
      end

      def response_body
        [::File.read(path)]
      end

      def self.find_first(options)
        asset_compiler = options[:asset_compiler]
        path_info = options[:path_info].chomp('/')

        return nil if asset_compiler.compiles?(path_info)

        asset_root = options[:asset_root]

        candidate_path_infos = []
        if !path_info.empty?
          candidate_path_infos << path_info
          candidate_path_infos << "#{path_info}#{DEFAULT_STATIC_EXTENSION}"
        end
        candidate_path_infos << "#{path_info}/index#{DEFAULT_STATIC_EXTENSION}"

        file_path = nil

        full_path_info = candidate_path_infos.find do |candidate_path_info|
          file_path = ::File.join(asset_root, candidate_path_info)
          readable_file?(file_path)
        end

        return nil if full_path_info.nil? || !has_static_extension?(full_path_info)

        include_gzipped = options[:include_gzipped]

        gzipped_file_path = "#{file_path}.gz"
        gzipped_file_present = readable_file?(gzipped_file_path)

        has_encoding_variants = gzipped_file_present

        init_options = {
          :path => file_path,
          :full_path_info => full_path_info,
          :has_encoding_variants => has_encoding_variants,
          :max_age_fallback => options[:max_age_fallback]
        }

        if include_gzipped && gzipped_file_present
          init_options[:path] = gzipped_file_path
          init_options[:is_gzipped] = true
        end

        return ServeableFile.new(init_options)
      end

      def self.has_static_extension?(path)
        Rack::Zippy.static_extensions.include? ::File.extname(path).slice(1..-1).to_s.downcase
      end

      def encoding_variants?
        return @has_encoding_variants
      end

      def gzipped?
        return @is_gzipped
      end

      def ==(other)
        return false if other.nil?
        return true if self.equal?(other)
        return self.class == other.class &&
          self.gzipped? == other.gzipped? &&
          self.encoding_variants? == other.encoding_variants? &&
          self.path == other.path &&
          self.full_path_info == other.full_path_info
      end
      alias_method :eql?, :==

        private

      # Old last-modified headers encourage caching via browser heuristics. Use it for year-long cached assets.
      CACHE_FRIENDLY_LAST_MODIFIED = 'Mon, 10 Jan 2005 10:00:00 GMT'

      SECONDS_IN = {
        :day => 24*60*60,
        :month => 31*(24*60*60),
        :year => 365*(24*60*60)
      }.freeze

      DEFAULT_STATIC_EXTENSION = '.html'.freeze

      def self.readable_file?(file_path)
        return ::File.file?(file_path) && ::File.readable?(file_path)
      end

    end
  end
end


lib/rack-zippy.rb
require 'rack-zippy/version'
require 'rack-zippy/configuration'
require 'rack-zippy/railtie' if defined?(Rails)
require 'action_controller'

module Rack
  module Zippy
    extend Configuration

    define_setting :static_extensions, %w(css js html htm txt ico png jpg jpeg gif pdf svg zip gz eps psd ai woff woff2 ttf eot otf swf)

    ASSETS_SUBDIR_REGEX = /\A\/assets(?:\/|\z)/

    class AssetServer

      attr_reader :static_middleware

      # @param app [#call(env)] the Rack app
      # @param path [String] the path to the public directory, usually where favicon.ico lives
      # @param max_age_fallback [Fixnum] time for Cache-Control header. Defaults to 1 day (in seconds).
      def initialize(app, path, max_age_fallback: :day)
        assert_path_valid path

        cache_control = cache_control(max_age_fallback)

        @app = app
        blank_app = ->(env) { }
        @static_middleware = ::ActionDispatch::Static.new(blank_app, path, cache_control)
      end

      def call(env)
        path_info = env[PATH_INFO]
        return not_found_response if illegal_path?(path_info)

        if try_static?(path_info)
          static_response = static_middleware.call(env)
        end

        if static_response
          after_static_responds(env, static_response)
        else
          @app.call(env)
        end
      end

      private

      ILLEGAL_PATH_REGEX = /(\/\.\.?)/

      BLANK_PATH_MESSAGE = 'Please specify non-blank path when initializing rack-zippy middleware ' +
        '(path leads to your public directory, often the one with favicon.ico in it)'

      SECONDS_IN = {
        day: 24*60*60,
        month: 31*(24*60*60),
        year: 365*(24*60*60)
      }.freeze

      # Old last-modified headers encourage caching via browser heuristics. Use it for year-long cached assets.
      CACHE_FRIENDLY_LAST_MODIFIED = 'Mon, 10 Jan 2005 10:00:00 GMT'.freeze

      FAVICON_PATH = '/favicon.ico'.freeze
      PATH_INFO = 'PATH_INFO'.freeze
      CACHE_CONTROL = 'Cache-Control'.freeze
      LAST_MODIFIED = 'Last-Modified'.freeze

      def assert_path_valid(path)
        raise ArgumentError, BLANK_PATH_MESSAGE if path.blank?
      end

      def cache_control(max_age_fallback)
        max_age_fallback = calc_max_age_fallback(max_age_fallback)
        "public, max-age=#{max_age_fallback}"
      end

      def calc_max_age_fallback(max_age_fallback)
        max_age_fallback.is_a?(Symbol) ? SECONDS_IN.fetch(max_age_fallback) : max_age_fallback
      end

      def illegal_path?(path)
        path =~ ILLEGAL_PATH_REGEX
      end

      def not_found_response
        [404, {}, ['Not Found']]
      end

      def try_static?(path)
        extension = extension(path)
        try_default_extension = extension.nil?
        try_default_extension || static_extension?(extension)
      end

      def extension(path)
        ext = ::File.extname(path).slice(1..-1)
        ext.downcase! if ext
        ext
      end

      def static_extension?(extension)
        Rack::Zippy.static_extensions.include? extension
      end

      def after_static_responds(env, static_response)
        path = ::Rack::Utils.unescape(env[PATH_INFO])
        headers = static_response[1]
        modify_headers(path, headers)
        static_response
      end

      def modify_headers(path, headers)
        case path
        when ASSETS_SUBDIR_REGEX
          lifetime_in_secs = SECONDS_IN[:year]
          last_modified = CACHE_FRIENDLY_LAST_MODIFIED
        when FAVICON_PATH
          lifetime_in_secs = SECONDS_IN[:month]
          last_modified = CACHE_FRIENDLY_LAST_MODIFIED
        end

        headers[CACHE_CONTROL] = "public, max-age=#{lifetime_in_secs}" if lifetime_in_secs

        if last_modified
          headers[LAST_MODIFIED] = last_modified
        else
          headers.delete(LAST_MODIFIED)
        end
      end

    end
  end
end
